grakn-graph/src/main/java/ai/grakn/graph/internal/RelationTypeImpl.java

```diff
@@ -80,6 +76,13 @@ public Relation addRelation() {
     public RelationType hasRole(RoleType roleType) {
         checkTypeMutation();
         putEdge(roleType, Schema.EdgeLabel.HAS_ROLE);
+
+        //Cache the Role internally
+        cachedHasRoles.get().add(roleType);
```

Comment:
It kind of has to be done that way I think. If you don't update the cache fully with all the roles then you simply adding the new role to the cache and that will result in only that role ever being returned when calling `hasRoles()`. This could be improved by tracking internally in `Cache` if the db has been read before but because cache is monitoring Sets and single elements the implementation will not be so simple.